File: Building a Web App with ASP.NET Core, MVC 6, EF Core, and Angular/12-asp.net-core-deployment/TheWorld/wwwroot/js/tripsController.js
// tripsController.js
(function () {

  "use strict";

  // Getting the existing module
  angular.module("app-trips")
    .controller("tripsController", tripsController);

  function tripsController($http) {

    var vm = this;

    vm.trips = [];

    vm.newTrip = {};

    vm.errorMessage = "";
    vm.isBusy = true;

    $http.get("/api/trips")
      .then(function (response) {
        // Success
        angular.copy(response.data, vm.trips);
      }, function (error) {
        // Failure
        vm.errorMessage = "Failed to load data: " + error;
      })
      .finally(function () {
        vm.isBusy = false;
      });

    vm.addTrip = function () {

      vm.isBusy = true;
      vm.errorMessage = "";

      $http.post("/api/trips", vm.newTrip)
        .then(function (response) {
          // success
          vm.trips.push(response.data);
          vm.newTrip = {};
        }, function () {
          // failure
          vm.errorMessage = "Failed to save new trip";
        })
        .finally(function () {
          vm.isBusy = false;
        });

    };



  }

})();

File: Building a Web App with ASP.NET Core, MVC 6, EF Core, and Angular/08-creating-the-api/TheWorld/Controllers/Api/TripsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AutoMapper;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using TheWorld.Models;
using TheWorld.ViewModels;

namespace TheWorld.Controllers.Api
{
  [Route("api/trips")]
  public class TripsController : Controller
  {
    private ILogger<TripsController> _logger;
    private IWorldRepository _repository;

    public TripsController(IWorldRepository repository, ILogger<TripsController> logger)
    {
      _repository = repository;
      _logger = logger;
    }

    [HttpGet("")]
    public IActionResult Get()
    {
      try
      {
        var results = _repository.GetAllTrips();

        return Ok(Mapper.Map<IEnumerable<TripViewModel>>(results));
      }
      catch (Exception ex)
      {
        _logger.LogError($"Failed to get All Trips: {ex}");

        return BadRequest("Error occurred");
      }
    }

    [HttpPost("")]
    public async Task<IActionResult> Post([FromBody]TripViewModel theTrip)
    {
      if (ModelState.IsValid)
      {
        // Save to the Database
        var newTrip = Mapper.Map<Trip>(theTrip);
        _repository.AddTrip(newTrip);

        if (await _repository.SaveChangesAsync())
        {
          return Created($"api/trips/{theTrip.Name}", Mapper.Map<TripViewModel>(newTrip));
        }
      }

      return BadRequest("Failed to save the trip");
    }
  }
}
